[3기 Tiaz0128] 자바스크립트 계산기 미션

tiaz0128 opened the pull request:
## 🎯 기능 요구사항
 
- [x] 2개의 숫자에 대해 덧셈이 가능하다.
- [x]  2개의 숫자에 대해 뺄셈이 가능하다.
- [x]  2개의 숫자에 대해 곱셈이 가능하다.
- [x]  2개의 숫자에 대해 나눗셈이 가능하다.
- [x]  AC(All Clear)버튼을 누르면 0으로 초기화 한다.
- [x]  숫자는 한번에 최대 3자리 수까지 입력 가능하다.
- [x]  계산 결과를 표현할 때 소수점 이하는 버림한다.

<br/>

## 🌏 데모 사이트
[링크](https://clever-yonath-b39e18.netlify.app/)

<br/>

## 📊 구현 기능 상세
- `App.js` - `Calculator.js` - `util.js` 로 구성
- `App.js` 에서 컴포넌트 계산기(Calculator)를 관리하는 형태로 구현
- 계산기의 버튼의 종류에 따라서 이벤트 처리 핸들러 함수등록 `onDigitsClick` `onOperationsClick`, `onAcClick`

```
숫자1 연산자 숫자2

123 + 100
```
- 위의 형태로 수식을 구성 할 수 있게 입력을 제한
  - 연산자 뒤에 바로 연산자가 올수 없게 제한.
  - 연산자는 첫번째 숫자가 음수인것을 표현하는 부호를 제외하고 연산자는 하나만 사용 가능
- 0으로 시작하는 숫자 표시 처리
- 첫번째 숫자가 음수인 경우에도 계산 할 수 있게 처리
- 무한대 또는 NaN 인 경우
  - 숫자 입력시 바로 알림창 없이 바로 다시 수식입력 가능
  - 연산자 입력시 알림창으로 안내

<br/>

## 🙉 아직 구현하지 못한 기능

- [x] 두번째 숫자가 0으로만 입력되는 경우 `000` 이 아닌 `0` 으로 표현하게 처리
- [x] NaN 처리
- [x] 무한대 처리
- [x] NaN, 무한대인 경우에 다시 처음부터 입력하게 유도
- [x] 리팩터링 작업 - 메세지, 리터럴 숫자 상수화. 변수, 함수명 변경하기. 객체 분할 활용하기
- [ ] POM(Page Object Modle) 형태로 테스트 코드 구성

<br/>

## ✏ 구현시 공부한 내용
- `cypress` 가 `Mocha`, `chai` 기반이라 3가지를 공부.
- `commands.js` 를 사용한 커스텀 명령어 작성방법.

<br/>

## 🤔 구현시 어려웠던 부분
- 테스트 케이스를 작성할때 어떤것을 성공으로 판단 할것인지, 기준을 어떻게 잡아야하는지 (시각적인 것? 데이터?)
- 만들어 둔 케이스 자체가 잘 못됀 경우, 테스트 케이스를 빈틈 없이 온전하게 잘 작성하는 방법.
- `should`, `expect`, `assert` 으로 검증하는 문법을 이해하는데 시간이 걸림.
- 리팩터링 작업과 기능 구현을 분리하면서 작업하기.

<br/>

## 🎁 참고한 링크
- [Node.js 테스트 프레임워크 Mocha](https://heropy.blog/2018/03/16/mocha/)
- [Mocha Tutorial - Step 01. Hello World!](https://kdydesign.github.io/2017/06/15/Mocha-step-01/)
- [Part 1: Cypress Introduction | Web Automation using Cypress](https://www.youtube.com/watch?v=yM38NmF-LC4&list=PLUDwpEzHYYLu4jKg-rNSKH3aJeBinlPXp&index=1)
- [Cypress Beginner Tutorial 1 | Introduction](https://www.youtube.com/watch?v=CYcdT-tOvB0&list=PLhW3qG5bs-L9LTfxZ5LEBiM1WFfvX3dJo)
- [Cypress Overview | Part 1](https://www.youtube.com/watch?v=IMx7Ng-T2mw&list=PLzDWIPKHyNmK9NX9_ng2IdrkEr8L4WwB0&index=1)

tiaz0128 committed:
계산기 프로젝트 컴포넌트화

tiaz0128 committed:
Cypress 테스트 세팅

tiaz0128 committed:
test : .digit 을 누르면 숫자가 h1#total에 들어간다.

tiaz0128 committed:
기능추가 : 숫자키를 누르면 숫자가 #total에 들어간다

tiaz0128 committed:
테스트 수정 : .digit 을 누르면 숫자가 h1#total에 들어간다.

tiaz0128 committed:
테스트 추가 : 숫자가 0인 경우 0을 계속 눌러도 0으로 표시한다.

tiaz0128 committed:
기능 추가 : 숫자가 0인 경우 0을 눌러도 0으로 표시

tiaz0128 committed:
테스트 수정 : contains로 변경 테스트명 변경

- should 명령어에서 contains로 변경
- 초기값 숫자가 0인 경우 0을 계속 눌러도 0으로 표시한다.

tiaz0128 committed:
테스트 추가 : 초기값 숫자가 0인 경우 0이 아닌 숫자를 누르면 제일 앞에 0을 지우고 변경하기

tiaz0128 committed:
기능 추가 : 초기값 숫자가 0인 경우 0이 아닌 숫자를 누르면 제일 앞에 0을 지우고 변경하기

tiaz0128 committed:
테스트 추가 : 앞에 숫자가 하나 있고 /X-+연산자 버튼을 누르면 연산자가 표시 되게. =연산자 제외

tiaz0128 committed:
기능 추가 : 앞에 숫자가 하나 있고 /X-+연산자 버튼을 누르면 연산자가 표시 되게. =연산자 제외

tiaz0128 committed:
테스트 수정 : 앞에 숫자가 하나 있고 /X-+연산자 버튼을 누르면 연산자가 표시 되게. =연산자 제외

tiaz0128 committed:
테스트 수정 : = 연산자를 경우 표시하지 않는다.

tiaz0128 committed:
기능 추가 : = 연산자를 경우 표시하지 않는다

tiaz0128 committed:
기능 요구사항 - 테스트 추가 : 2개의 숫자에 대해 덧셈이 가능

tiaz0128 committed:
기능 요구사항 - 기능 구현 : 2개의 숫자에 대해 덧셈이 가능

tiaz0128 committed:
테스트 리팩터링

- inputNumber
- inputRepeatNumber
- inputRandomNumericalExpression
- inputIgnoreTargetNumber
- inputOperator
- inputIgnoreTargetOperator

tiaz0128 committed:
기능 요구사항 - 테스트 추가 : 2개의 숫자에 대해 뺄셈이 가능

tiaz0128 committed:
기능 요구사항 - 기능 구현 : 2개의 숫자에 대해 뺄셈이 가능

tiaz0128 committed:
기능 요구사항 - 테스트 추가 : 2개의 숫자에 대해 곱셈이 가능

tiaz0128 committed:
기능 요구사항 - 기능 구현 : 2개의 숫자에 대해 곱셈이 가능

tiaz0128 committed:
기능 요구사항 - 테스트 추가 : 2개의 숫자에 대해 나눗셈이 가능

tiaz0128 committed:
기능 요구사항 - 기능 구현 : 2개의 숫자에 대해 나눗셈이 가능

tiaz0128 committed:
기능 요구사항 - 테스트 수정 : 2개의 숫자에 대해 나눗셈이 가능

- util.js parseNumber의 로직을 parseInt 에서 parserFloat 으로 변경
- commands.js calculateNumericalExpression 로직 수정

tiaz0128 committed:
기능 요구사항 - 기능 수정 : 2개의 숫자에 대해 나눗셈이 가능

- util.js parseNumber 로직을 parserFloat 으로 변경

tiaz0128 committed:
테스트 리팩터링

- cy.get('#total').contains('0')
- cy.get('#total').contains(/^0$/gim) 로 변경

tiaz0128 committed:
기능 요구사항 - 테스트 추가 : AC(All Clear)버튼을 누르면 0으로 초기화

tiaz0128 committed:
기능 요구사항 - 기능 구현 : AC(All Clear)버튼을 누르면 0으로 초기화

tiaz0128 committed:
기능 요구사항 - 테스트 수정 : AC(All Clear)버튼을 누르면 0으로 초기화

- AC 이후 비교 결과값 문자열 비교로 변경

tiaz0128 committed:
기능 요구사항 - 테스트 추가 : 숫자는 한번에 최대 3자리 수까지 입력 가능

tiaz0128 committed:
기능 요구사항 - 기능 구현 : 숫자는 한번에 최대 3자리 수까지 입력 가능

tiaz0128 committed:
테스트 오류 수정

- targetOperator 파라미터 없을 경우 랜덤한 연산자 기능 오류 수정
- 연산자 리스트 operators 배열 추가
- inputIgnoreTargetOperator 에서 operators 사용하게 변경

tiaz0128 committed:
테스트 추가 : 연산자 연속으로 오는 경우 경고창

tiaz0128 committed:
기능 구현 : 연산자 연속으로 오는 경우 경고창

tiaz0128 committed:
테스트 수정 : 연산자 연속으로 오는 경우 경고창

tiaz0128 committed:
App.js 리팩터링

- this.state.totalNumber + e.target.innerText 변수화
- util.js isDuplicateOperator 추가

tiaz0128 committed:
테스트 추가 : 연산자는 젤 앞에 음수 기호 빼고 한개

tiaz0128 committed:
기능 구현 : 연산자는 젤 앞에 음수 기호 빼고 한개

tiaz0128 committed:
App.js util.js 리팩터링

- 연산자는 연속해서 입력 불가 기존 함수명 변경
- isDuplicateOperator -> isContinuousOperator 변경
- 연산자 갯수 확인하는 isDuplicateOperator 함수 추가

tiaz0128 committed:
테스트 추가 : 앞의 숫자가 음수일때 계산 가능 여부 확인

tiaz0128 committed:
테스트 수정 : parseNumericalExpression 로직 수정

- util.js 에 parseNumericalExpression 로직 문제 발견
- 음수 표현에 쓰는 - 기호는 제외하고 연산자만으로 연산하게 변경

tiaz0128 committed:
기능 수정 : 앞의 숫자가 음수일때 계산 가능하게 로직 변경

- 앞의 숫자가 음수일떄 계산 불가한 문제 수정

tiaz0128 committed:
테스트 수정 - commands.js 수정

- 매개변수 디폴값 undefined 삭제
- inputRepeatNumber 첫번째 매개변수(자릿수) 랜덤한 디폴트값 추가
-

tiaz0128 committed:
테스트 리팩터링

- 테스트에서 숫자 입력이 무작위 숫자 1 ~ 4 자리 입력하게 변경
- 무작위 수식이 들어갈 수 있게 변경

tiaz0128 committed:
기능 수정 : util.js isRightNumericalExpression 함수 수정

- 정규표현식 수정

tiaz0128 committed:
테스트 추가 : 수식이 완성하지 않고 =연산자 누르면 경고

tiaz0128 committed:
기능 추가 : 수식이 완성하지 않고 =연산자 누르면 경고

- util.js calculateNumericalExpression 함수 추가
- 수식이 완성되면 계산한 결과값 리턴
- 수식이 완성되지 않으면 경고창

tiaz0128 committed:
테스트 리팩터링

- cy.on 으로 window:alert 메시지 확인하는 방식 추가

tiaz0128 committed:
기능 요구사항 - 테스트 추가 : 계산 결과 소수점 이하 버림

tiaz0128 committed:
테스트 리팩터링 - 테스트 케이스 위치 변경

- 기능 요구사항 하단에 배치

tiaz0128 committed:
기능 요구사항 - 기능 구현 : 계산 결과 소수점 이하 버림

- util.js 에서 parseNumber 로직을 parseFloat 에서 parseInt 변경

tiaz0128 committed:
App.js 리팩터링

- 필요없는 state 삭제

yujo11 commented on cypress/integration/calculatorTest.spec.js:
```diff
@@ -0,0 +1,152 @@
+describe('계산기 테스트', () => {
+  beforeEach(() => {
+    cy.visit('http://127.0.0.1:5500/index.html')
+    cy.get('#total').contains(/^0$/gim)
```
오 이건 어떤 역할을 해주는 정규식인가요?

tiaz0128 replied:
처음 시작값이 0으로 세팅되어 있는지 확인하는 부분입니다! 😀

yujo11 commented on cypress/integration/calculatorTest.spec.js:
```diff
@@ -0,0 +1,152 @@
+  it('앞에 숫자가 하나 있고 /X-+연산자 버튼을 누르면 연산자가 표시 되게. =연산자 제외', () => {
+    cy.inputRepeatNumber()
+    cy.inputIgnoreTargetOperator('=')
+    cy.get('#total').contains(/^-?[0-9]+[+/X-]$/gim)
```
정규식을 잘 활용하시는군요 멋지네요!👍👍👍

yujo11 commented on cypress/support/util.js:
```diff
@@ -0,0 +1,31 @@
+}
+
+const parseNumericalExpression = (rawStr) => {
+  const reg = /^(-?[0-9]+)([+/X-])([0-9]+)$/gim
+
+  const num1 = rawStr.replace(reg, '$1')
+  const operator = rawStr.replace(reg, '$2')
+  const num2 = rawStr.replace(reg, '$3')
```
오... 정규식을 이렇게 활용할 수도 있군요 배워갑니다!👍👍👍

yujo11 commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,75 @@
+
+const operators = ['+', '-', 'X', '/', '=']
+
+Cypress.Commands.add('inputNumber', (targetNumber) => {
```
commands를 이렇게 활용하는거군요ㅎㅎ 배워갑니다!👍👍👍

tiaz0128 committed:
테스트 추가 : 두번째 숫자가 0이 먼저 오는 경우 표시

tiaz0128 committed:
기능 구현 : 두번째 숫자가 0이 먼저 오는 경우 표시

tiaz0128 committed:
util.js 리팩터링 : 정규표현식 변수명 통일, 복수형 변수명으로 변경

tiaz0128 committed:
calculatorTest.spec.js 리팩터링 : 테스트 분리

- 기능 요구사항 / 입력 테스트 분리
- 주석 정리

tiaz0128 committed:
테스트 추가 : 0/0 인 경우 NaN 처리 - 팝업, 이전값으로 되돌리기?

tiaz0128 committed:
commands.js 리팩터링 : OPERATOR, NUMBERS 상수화

tiaz0128 committed:
util.js parseNumericalExpression 리턴값 정수형에서 문자열로 변환

tiaz0128 committed:
commands.js 에서 parseInt 부분 제외

- 처음에는 문자열로 시작했다가 계산하면서 정수형으로 변경
- 이 때문에 처음과 비교하는 부분에서 제대로 비교 연산이 불가
- parseNumericalExpression 리턴값을 toString으로 문자열로 변환
- 문자열로 검사하게 로직 변경

skid901 commented on src/js/component/App.js:
```diff
@@ -0,0 +1,53 @@
+    this.calculator = new Calculator({
+      $calculatorDom: this.$calculatorDom,
+      initState: this.state,
+      onDigitsClick: (e) => {
+        let newStr = this.state.totalNumber + e.target.innerText
+
+        if (this.state.totalNumber === '0' && e.target.innerText === '0') return
+
+        newStr = covertZeroZero(newStr)
+
+        if (isLimitDigit(newStr)) return
+
+        this.setState({ totalNumber: this.state.totalNumber === '0' ? e.target.innerText : newStr })
+      },
```
```js
onDigitsClick: ({ target }) => { /** ... */ }
// or
onDigitsClick: ({ target: { innerText } }) => { /** ... */ }
```
위와 같이 디스트럭쳐링을 사용해 보시는 것도 좋을 것 같아요

skid901 commented on src/js/component/App.js:
```diff
@@ -0,0 +1,53 @@
+    this.calculator = new Calculator({
+      $calculatorDom: this.$calculatorDom,
+      initState: this.state,
+      onDigitsClick: (e) => {
+        let newStr = this.state.totalNumber + e.target.innerText
+
+        if (this.state.totalNumber === '0' && e.target.innerText === '0') return
+
+        newStr = covertZeroZero(newStr)
+
+        if (isLimitDigit(newStr)) return
+
+        this.setState({ totalNumber: this.state.totalNumber === '0' ? e.target.innerText : newStr })
+      },
```
```js
onDigitsClick: ({ target }) => { /** ... */ }
// or
onDigitsClick: ({ target: { innerText } }) => { /** ... */ }
```
위와 같이 디스트럭쳐링을 사용해 보시는 것도 좋을 것 같아요

tiaz0128 replied:
넵 좋은 생각입니다! 반영 해보겠습니다! 😀

skid901 commented on src/js/component/Calculator.js:
```diff
@@ -0,0 +1,32 @@
+    this.render()
+  }
+
+  this.setState = (newState) => {
+    this.state = newState
+    this.render()
+  }
```
```js
this.setState = (newState) => {
  this.state = { ...this.state, ...newState };
  this.render();
};
```
`this.state` 값이 `Object` 타입이면
상태를 변경하는 로직에 디스트럭쳐링을 활용해
조금 더 편리하게 사용할 수 있을 것 같아요

tiaz0128 replied:
감사합니다! 분할하는게 써야지하면서 계속 까먹었는데 
요번 스터디 동안 열심히 배워가겠습니다! 앞으로도 잘 부탁드립다! 😁

skid901 commented on src/js/component/util.js:
```diff
@@ -0,0 +1,78 @@
+  return /^-?[0-9]+([+/X-])[0-9]+$/gim.test(rawStr)
+}
+
+const parseNumericalExpression = (rawStr) => {
+  const reg = /^(-?[0-9]+)([+/X-])([0-9]+)$/gim
+
+  const num1 = rawStr.replace(reg, '$1')
+  const operator = rawStr.replace(reg, '$2')
+  const num2 = rawStr.replace(reg, '$3')
+
+  return parseNumber(calculate(num1, operator, num2)).toString()
+}
```
정규표현식을 정말 잘 사용하시는 것 같아요 ㄷㄷ
정규표현식 활용법 배워갑니다 ^^7

skid901 reviewed (commented):
안녕하세요, @Tiaz0128 님.
제 코드를 리뷰해 주셔서 감사합니다 ㅎㅎ
스터디 기간 동안 잘 부탁드립니다 ^^7

tiaz0128 committed:
테스트 수정 - commands.js 수정

- parseInt 에서 pareFloat 으로 변경
- 나누기의 경우 Math.floor 로 소수점 처리

tiaz0128 committed:
component/*.js 리팩터링

- 변수명 통일 및 변경
- 조건문과 return 시 불필요한 연산 삭제
- 검증 함수에서 alert 분리해서 리턴값에서 호출하게 변경
- 기존에 정수값에서 나눗셈 계산을 실수값으로 변경 후 처리
- 불필요한 함수 삭제 util.js calculateNumericalExpression

tiaz0128 committed:
util.js 리팩터링

- Null 인 경우 Nullish coalescing operator 를 이용
- 리턴값과 조거문을 줄이는데 사용
- https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

tiaz0128 committed:
App.js 리팩터링 - 메시지 상수화

- App.js 리터럴 메시지 분리
- 메시지 상수 관리하는 constant.js 파일 추가

tiaz0128 committed:
테스트 수정 : 0/0 인 경우 NaN 처리 - 다시 수식 입력가능

tiaz0128 committed:
commands.js 리팩터링 : 필요없는 조건 제거

tiaz0128 committed:
기능 구현 : 0/0 인 경우 NaN 처리 - 다시 수식 입력가능

tiaz0128 committed:
테스트 수정 : 0/0 인 경우 NaN 처리 - 다시 수식 입력가능

- 검증 부분 expect 로 다시 구현

tiaz0128 committed:
테스트 추가 : 무한대 처리

- 양수  1/0 인 경우 Infinity 처리 - 다시 수식 입력가능
- 음수 -1/0 인 경우 -Infinity 처리 - 다시 수식 입력가능

tiaz0128 committed:
기능 추가 : 숫자가 아닌 경우에 연산자 입력하는 경우 팝업 알림

- 숫자가 아닌 상태에서 연산자 입력에서 하면 팝업 알림
- constant.js 에 RESTART_NUMERICAL_EXPRESSION 추가

tiaz0128 commented:
## 코드리뷰 반영 및 배운점
다른 분들의 코드와 PR 을 보면서 고친것들 배운것들을 적어봅니다!
- `??` 연산자 [Nullish coalescing operator](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator) 을 사용해서 null 인 경우 조금 더 효율적인 코딩 가능
- `e.target.innerText` 와 같은 경우 `(e)` 를 객체 분할을 통해서 `({ target: { innerText: operatorChar } })` 형태로 뽑아내기
- 메시지 리터럴 상수화와 객체 `Object.freeze` 활용

skid901 commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,76 @@
+// Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
+
+import { times, random } from 'lodash'
+import { parseNumericalExpression, parseNumber } from './util.js'
```
`cypress`에서 사용할 `util` 함수를 제공하는 방법이 있었네요!!
덕분에 배워갑니다 ^^7

skid901 commented on src/js/component/App.js:
```diff
@@ -0,0 +1,52 @@
+import { Calculator } from './Calculator.js'
+import { ERR_MSG } from './constant.js'
+import { isNotNumber, isLimitDigit, isContinuousOperator, isDuplicateOperator, isRightNumericalExpression, parseNumericalExpression, convertZeroZero } from './util.js'
```
해당 라인이 조금 길어서 한눈에 보기 어려울 것 같아요.
`prettier`를 사용해 보시면 어떠실까요?

tiaz0128 replied:
넵 추가하도록 하겠습니다!

pjh1995 commented on cypress/integration/calculatorTest.spec.js:
```diff
@@ -0,0 +1,223 @@
+    cy.inputRandomNumericalExpression('/')
+    cy.inputOperator('=')
+
+    cy.get('#total').should('not.contain.text', '.')
```
올림일 경우에도 소수점이 없으니 다르게 테스트 해보시면 어떨까요?

tiaz0128 replied:
그 부분을 생각하지 못했네요. 감사합니다!

skid901 commented on src/js/component/util.js:
```diff
@@ -0,0 +1,64 @@
+  return rawStr.match(reg) ? rawStr.replace(reg, '$1$2') : rawStr
+}
+
+const addClickEvent = (selector, func) => {
+  const targetDom = document.querySelector(selector)
+  targetDom.addEventListener('click', func)
+}
```
```js
if (targetDom) targetDom.addEventListener('click', func);
// or
targetDom?.addEventListener('click', func);
```
`targetDom`의 유효성을 확인하는 로직을 적용해보시면
더 안전하게 함수를 사용할 수 있을 것 같아요!!

tiaz0128 replied:
`?.` 도 처음 알게됐습니다! 감사합니다! 👍👍

pjh1995 commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,76 @@
+  targetNumber !== undefined ? cy.get('.digit').contains(targetNumber).click() : cy.get('.digit').contains(random(0, 9)).click()
+})
+
+Cypress.Commands.add('inputRepeatNumber', (maxDigit = random(1, 4), targetNumber) => {
```
테스트 할때마다 랜덤한 숫자가 나오게 되면 테스트 결과가 달라질 수 있지 않을까.. 싶습니다!

tiaz0128 replied:
넵. 저도 그부분을 조금 고민 중이긴 했습니다. 좀더 구체적인 테스트로 세분화시켜봐야겠습니다! 😁

pjh1995 commented on src/js/component/App.js:
```diff
@@ -0,0 +1,52 @@
+    this.calculator = new Calculator({
+      $calculatorDom,
+      initState: this.state,
+      onDigitsClick: ({ target: { innerText: digitChar } }) => {
```
클릭 이벤트 처리는 외부에서 하고, 계산하는 부분만 컴포넌트에 함수로 구현해두면 어떨까요?

skid901 commented on cypress/integration/calculatorTest.spec.js:
```diff
@@ -0,0 +1,223 @@
+
+  it('숫자를 누르면 h1#total에 들어간다.', () => {
+    cy.inputRepeatNumber()
+    cy.get('#total').contains(/[0-9]+/gim)
```
정규표현식을 아직 잘 모르는 개발자와 협업하시는 경우,
정규표현식 객체를 해당 정규표현식을 설명하는 이름을 가진 변수에 할당해서 사용해 보시면 어떠실까요?
개인적으로 정규식을 아직 잘 몰라서 문득 이러한 생각이 들었습니다 ^^;;;

tiaz0128 replied:
넵! 변수명이나 주석으로 설명할수 있게해두겠습니다!

pjh1995 commented on src/js/component/constant.js:
```diff
@@ -0,0 +1,6 @@
+export const ERR_MSG = {
+  MAX_DIGIT_THREE: '숫자는 세 자리까지만 입력 가능합니다!',
+  MAX_OPERATOR_ONE: '연산자는 한개만 가능',
+  CONTINUOUS_OPERATOR_ONE: '연산자는 연속해서 입력 불가',
+  RESTART_NUMERICAL_EXPRESSION: '처음부터 수식을 다시 입력 해주세요',
+}
```
메세지를 정리해두니 깔끔하네요 한수 배워갑니다 👍

skid901 reviewed (commented):
전에 리뷰드렸던 내용을 바로 반영하셔서 정말 놀랐습니다 ㄷㄷ
전에 한번 리뷰를 드려서 그런지 이번에는 다소 개인적인 의견이 많이 들어간 리뷰를 드린 것 같네요;;
부족하나마 도움이 되시길 바랍니다 ㅜㅜ

pjh1995 commented on src/js/component/App.js:
```diff
@@ -0,0 +1,52 @@
+
+        this.setState({ rawStr: newStr })
+      },
+      onOperationsClick: ({ target: { innerText: operatorChar } }) => {
```
해당 타겟이 .operation인지 체크하는 부분이 들어가면 좋을 것 같습니다 ㅎ

pjh1995 commented on src/js/component/App.js:
```diff
@@ -0,0 +1,52 @@
+        const rawStr = this.state.rawStr
+        if (rawStr === '0' || isNotNumber(rawStr)) return this.setState({ rawStr: digitChar })
+
+        const newStr = convertZeroZero(rawStr + digitChar)
```
어떤 역할을 하는지 살짝 모호한것 같습니다! 
네이밍을 구체적으로 바꿔보시면 어떨까요?

pjh1995 reviewed (commented):
수고하셨습니다 👍
